package org.elnix.sle.ui.settings

import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import io.github.elnix90.core.objects.FloatSettingObject
import io.github.elnix90.runtime.asState
import kotlinx.coroutines.launch
import org.elnix.sle.ui.components.DragonGroupScope
import org.elnix.sle.ui.components.SliderWithLabel

@Composable
fun DragonGroupScope.Setting(
	setting: FloatSettingObject,
	decimals: Int = 2,
	enabled: Boolean = true,
	customDesc: ((Float) -> String)? = null
) {
	val ctx = LocalContext.current
	val scope = rememberCoroutineScope()

	val state by setting.asState()

	var tempState by remember { mutableFloatStateOf(state) }

	LaunchedEffect(state) { tempState = state }

	SliderWithLabel(
		label = stringResource(setting.title!!),
		description = customDesc?.invoke(state) ?: stringResource(setting.description!!),
		value = tempState,
		valueRange = setting.allowedRange,
		enabled = enabled,
		decimals = decimals,
		resetEnabled = tempState != setting.default,
		onReset = { scope.launch { setting.reset(ctx) } },
		onDragStateChange = { scope.launch { setting.set(ctx, tempState) } }
	) { tempState = it }
}
